peak/features.py

```diff
@@ -0,0 +1,141 @@
+from collections import OrderedDict
+from hwtypes.adt import Product, Tuple
+import functools
+from ast_tools.passes import begin_rewrite, end_rewrite
+from ast_tools.passes import ssa, bool_to_bit, if_to_phi
+from ast_tools.stack import SymbolTable
+from hwtypes import SMTBit
+from peak.assembler import Assembler
+import magma
+
+def name_outputs(**outputs):
+    """Decorator meant to apply to any function to specify output types
+    The output type will be stored in fn._peak_outputs__
+    The input type will be stored in fn._peak_inputs_
+    Will verify that all the inputs have type annotations
+    Will also verify that the outputs of running fn will have the correct number of bits
+    """
+    peak_outputs = OrderedDict()
+    for oname, otype in outputs.items():
+        peak_outputs[oname] = otype
+    output_t = Product.from_fields("Output", peak_outputs)
+    def decorator(call_fn):
+        call_fn._output_t = output_t
+        return call_fn
+    return decorator
+
+def gen_input_t(call_fn):
+    try:
+        input_t = call_fn._input_t
+    except:
+        #construct input_t
+        arg_offset = 1 if call_fn.__name__ == "__call__" else 0
+        peak_inputs = OrderedDict()
+        num_inputs = call_fn.__code__.co_argcount
+        input_names = call_fn.__code__.co_varnames[arg_offset:num_inputs]
+        in_types = call_fn.__annotations__
```

Comment:
Should probably use inspect for all of this.  It will undoubtedly be more robust.

https://docs.python.org/3/library/inspect.html#inspect.signature
https://docs.python.org/3/library/inspect.html#inspect.getfullargspec

Reply:
Ill add this as a separate issue/PR. This is how name_outputs was working for the last year.

Reply:
Fair enough